### PR TITLE
2405: Add Hash2CryptoLib [Rebase & FF]

### PR DESCRIPTION
**Pull request by makubacki**
## Description

Adds a new library class for a library that produces the Hash 2 Service Binding
and Hash 2 protocols.

---

**SecurityPkg: Add Hash2CryptoLib**

Adds a new library instance that abstracts implementation of Hash 2
cryptography services. This abstraction is primarily added as a
library to allow a platform to link the code that produces the UEFI
Hash 2 Service Binding protocol and Hash 2 protocol to a module that
already links against the cryptography functionality to reduce
overall platform size.

---

**SecurityPkg/Hash2CryptoDxe: Use Hash2CryptoLib**

Updates the module to use the functionality moved from this module
to the DxeHash2CryptoLib implementation.

---

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

- Changes from previous release branch with no functional changes.

## Integration Instructions

If a module is already linked against hash crypto code, it can link
this library to produce the hash protocols saving space in the platform
by consolidating the hashing code to one location.

### CODE REVIEW

File: SecurityPkg/Include/Library/Hash2CryptoLib.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/** @file
+  This library abstracts installation of the Hash 2 Service Binding protocol.
+
+  Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (os-d) [inline SecurityPkg/Include/Library/Hash2CryptoLib.h]: Update the licenses here.

File: SecurityPkg/Include/Library/Hash2CryptoLib.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
+/** @file
+  This library abstracts installation of the Hash 2 Service Binding protocol.
+
+  Copyright (c) Microsoft Corporation. All rights reserved.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef _HASH2_CRYPTO_LIB_H_
+#define _HASH2_CRYPTO_LIB_H_
```
Comment (os-d) [inline SecurityPkg/Include/Library/Hash2CryptoLib.h]: nit: not standard style in these headers

File: SecurityPkg/Library/BaseHash2CryptoLibNull/BaseHash2CryptoLibNull.uni [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+// **/
+
+// MU_CHANGE - NEW FILE
+
+#string STR_MODULE_ABSTRACT             #language en-US "Null Instance of Hash 2 Crypto Library"
+
+#string STR_MODULE_DESCRIPTION          #language en-US "Null instance of Hash 2 Crypto Library."
```
Comment (os-d) [inline SecurityPkg/Library/BaseHash2CryptoLibNull/BaseHash2CryptoLibNull.uni:13]: I'm fine with this, but we seem inconsistent about whether we have uni files or not for files that don't actually have strings

**Review by os-d — APPROVED** [review]

File: SecurityPkg/Library/BaseHash2CryptoLibNull/BaseHash2CryptoLibNull.uni (hunk 3 above)
Comment (makubacki) [reply SecurityPkg/Library/BaseHash2CryptoLibNull/BaseHash2CryptoLibNull.uni:13]: I usually do include them in "core" code, often they have not been. I'll keep this file.

File: SecurityPkg/Include/Library/Hash2CryptoLib.h (hunk 1 above)
Comment (makubacki) [reply SecurityPkg/Include/Library/Hash2CryptoLib.h]: This was submitted earlier on, before forming opinions about changing these things. Will update.

File: SecurityPkg/Include/Library/Hash2CryptoLib.h (hunk 2 above)
Comment (makubacki) [reply SecurityPkg/Include/Library/Hash2CryptoLib.h]: This was submitted earlier on, before forming opinions about changing these things. Will update.

**Review by kenlautner — APPROVED** [review]

**Pull request merged**